pallets/gear-builtin-actor/src/lib.rs

```diff
@@ -0,0 +1,394 @@
+    fn from(err: BuiltinActorError) -> Self {
+        match err {
+            BuiltinActorError::InsufficientGas => SimpleExecutionError::RanOutOfGas,
+            BuiltinActorError::UnknownMessageType => SimpleExecutionError::UserspacePanic,
+            BuiltinActorError::UnknownBuiltinId => SimpleExecutionError::BackendError,
+        }
+    }
+}
+
+pub type BuiltinResult<P> = Result<P, BuiltinActorError>;
+
+/// A trait representing an interface of a builtin actor that can receive a message
+/// and produce a set of outputs that can then be converted into a reply message.
+pub trait BuiltinActor<Message: Dispatchable, Gas: Zero> {
+    /// Handles a message and returns a result and the actual gas spent.
+    fn handle(message: &Message, gas_limit: Gas) -> (BuiltinResult<Message::Payload>, Gas);
+
+    fn get_ids(buffer: &mut Vec<BuiltinId>);
+}
+
+pub trait RegisteredBuiltinActor<M: Dispatchable, G: Zero>: BuiltinActor<M, G> {
+    /// The global unique ID of the trait implementer type.
+    const ID: BuiltinId;
+}
+
+// Assuming as many as 16 builtin actors for the meantime
+#[impl_for_tuples(16)]
+#[tuple_types_custom_trait_bound(RegisteredBuiltinActor<M, G>)]
+impl<M: Dispatchable, G: Zero> BuiltinActor<M, G> for Tuple {
+    fn handle(message: &M, gas_limit: G) -> (BuiltinResult<M::Payload>, G) {
+        for_tuples!(
+            #(
+                if (Tuple::ID == message.destination()) {
+                    return Tuple::handle(message, gas_limit);
+                }
+            )*
+        );
+        (Err(BuiltinActorError::UnknownBuiltinId), Zero::zero())
+    }
+
+    fn get_ids(buffer: &mut Vec<BuiltinId>) {
+        for_tuples!(
+            #(
+                buffer.push(Tuple::ID);
+            )*
+        );
+    }
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use frame_support::{pallet_prelude::*, traits::Get, PalletId};
+    use frame_system::pallet_prelude::*;
+
+    // This pallet doesn't define a storage version because it doesn't use any storage
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The builtin actor type.
+        type BuiltinActor: BuiltinActor<SimpleBuiltinMessage, u64>;
+
+        /// Weight cost incurred by builtin actors calls.
+        type WeightInfo: WeightInfo;
+
+        /// The builtin actor pallet id, used for deriving unique actors ids.
+        #[pallet::constant]
+        type PalletId: Get<PalletId>;
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    impl<T: Config> Pallet<T> {
+        /// Generate an `actor_id` given a builtin ID.
+        ///
+        ///
+        /// This does computations, therefore we should seek to cache the value at the time of
+        /// a builtin actor registration.
+        pub fn generate_actor_id(builtin_id: BuiltinId) -> ProgramId {
```

Comment:
we can do similar trick with that is used for vouchers hash(["bultin", idx]) so we will avoid using pallet id that could change from chain to chain based on gear

Reply:
and avoid adding rpc api

Reply:
Agree with the first comment. Still believe having an rpc is useful to make sure the respective `ProgramId` address is always up-to-date for a builtin id.